File: cpp/exercises/practice/unit_test_5cc94226_99e9_4c4a_9444_1936e3640000/task_scheduler/.meta/task_scheduler.cpp
#include "task_scheduler.h"
#include <vector>
#include <queue>
#include <functional>
#include <algorithm>

namespace task_scheduler {

int min_total_penalty(const std::vector<int>& duration,
                      const std::vector<int>& deadline,
                      const std::vector<std::vector<int>>& dependencies) {
    int n = duration.size();
    // Build graph: for each task, we need out_edges to children
    std::vector<std::vector<int>> graph(n);
    // Compute in-degree for each task. The "dependencies" vector for task i lists tasks that must precede i.
    std::vector<int> indegree(n, 0);
    for (int i = 0; i < n; ++i) {
        for (int pre : dependencies[i]) {
            // pre -> i edge
            graph[pre].push_back(i);
            indegree[i]++;
        }
    }
    
    // Priority queue to select the next available task.
    // We prioritize tasks with the earliest deadline.
    auto cmp = [&deadline](int left, int right) {
        return deadline[left] > deadline[right];
    };
    std::priority_queue<int, std::vector<int>, decltype(cmp)> pq(cmp);
    
    // Insert tasks with no dependencies (in-degree 0).
    for (int i = 0; i < n; ++i) {
        if (indegree[i] == 0) {
            pq.push(i);
        }
    }
    
    // Greedy scheduling: always pick the available task with the smallest deadline.
    long long current_time = 0;
    long long total_penalty = 0;
    
    while (!pq.empty()) {
        int task = pq.top();
        pq.pop();
        
        current_time += duration[task];
        if (current_time > deadline[task]) {
            total_penalty += (current_time - deadline[task]);
        }
        
        // Reduce the in-degree for tasks that depend on the current task.
        for (int next_task : graph[task]) {
            indegree[next_task]--;
            if (indegree[next_task] == 0) {
                pq.push(next_task);
            }
        }
    }
    
    return static_cast<int>(total_penalty);
}

}  // namespace task_scheduler

File: java/exercises/practice/unit_test_72dc79fd_3041_4d6c_af45_0115ba7413c2/task_scheduler/.meta/src/main/java/TaskScheduler.java
import java.util.*;

public class TaskScheduler {

    // Global variable for backtracking optimum result
    private static long bestTotalLateness;

    public static int minTotalLateness(List<Task> tasks) {
        int n = tasks.size();
        // Build task map by id for quick access.
        Map<Integer, Task> taskMap = new HashMap<>();
        for (Task t : tasks) {
            taskMap.put(t.id, t);
        }

        // Build graph: for each task, add to successors of its dependencies.
        // Also, compute in-degree for each task.
        Map<Integer, Integer> inDegree = new HashMap<>();
        for (Task t : tasks) {
            inDegree.put(t.id, t.dependencies.size());
        }
        for (Task t : tasks) {
            for (int dep : t.dependencies) {
                Task depTask = taskMap.get(dep);
                if (depTask != null) {
                    depTask.successors.add(t);
                }
            }
        }

        // For small number of tasks, use exhaustive backtracking to guarantee optimality.
        if(n <= 15) {
            bestTotalLateness = Long.MAX_VALUE;
            // Initialize available tasks: tasks with inDegree zero.
            List<Task> available = new ArrayList<>();
            for (Task t : tasks) {
                if (inDegree.get(t.id) == 0) {
                    available.add(t);
                }
            }
            // Make a copy of inDegree map for backtracking.
            backtrack(0, 0, inDegree, available, taskMap);
            return (int) bestTotalLateness;
        } else {
            // For larger set of tasks, use a greedy heuristic:
            // At each step, select from available tasks (in-degree 0) the task with earliest deadline.
            PriorityQueue<Task> queue = new PriorityQueue<>(Comparator.comparingInt(t -> t.deadline));
            for (Task t : tasks) {
                if (inDegree.get(t.id) == 0) {
                    queue.offer(t);
                }
            }
            long currentTime = 0;
            long totalLateness = 0;
            while (!queue.isEmpty()) {
                Task curr = queue.poll();
                currentTime += curr.duration;
                totalLateness += Math.max(0, currentTime - curr.deadline);

                // For each successor, reduce in-degree and add if becomes 0.
                for (Task succ : curr.successors) {
                    int deg = inDegree.get(succ.id) - 1;
                    inDegree.put(succ.id, deg);
                    if (deg == 0) {
                        queue.offer(succ);
                    }
                }
            }
            return (int) totalLateness;
        }
    }

    // Backtracking for optimal scheduling for small number of tasks.
    private static void backtrack(long currentTime, long currentLate, Map<Integer, Integer> inDegree, List<Task> available, Map<Integer, Task> taskMap) {
        // If no tasks are left, update best result.
        if (available.isEmpty()) {
            // Check if all tasks have been scheduled.
            boolean allScheduled = true;
            for (int deg : inDegree.values()) {
                if (deg != -1) { // We mark scheduled tasks with -1.
                    allScheduled = false;
                    break;
                }
            }
            if (allScheduled) {
                bestTotalLateness = Math.min(bestTotalLateness, currentLate);
            }
            return;
        }

        // Prune if current lateness already exceeds best found.
        if (currentLate >= bestTotalLateness) {
            return;
        }

        // Try each available task in turn.
        // To avoid side-effects, make a copy of available list for each recursive call.
        for (int i = 0; i < available.size(); i++) {
            Task task = available.get(i);

            // Prepare new available list for next recursion.
            List<Task> nextAvailable = new ArrayList<>(available);
            nextAvailable.remove(i);

            // Save state changes: copy inDegree map.
            Map<Integer, Integer> newInDegree = new HashMap<>(inDegree);
            // Mark current task as scheduled.
            newInDegree.put(task.id, -1);

            long finishTime = currentTime + task.duration;
            long lateness = Math.max(0, finishTime - task.deadline);
            long nextLate = currentLate + lateness;

            // For each successor, decrement in-degree and if becomes 0, add to available.
            for (Task succ : task.successors) {
                if (newInDegree.get(succ.id) != -1) { // if not already scheduled
                    int newDeg = newInDegree.get(succ.id) - 1;
                    newInDegree.put(succ.id, newDeg);
                    if (newDeg == 0) {
                        nextAvailable.add(succ);
                    }
                }
            }
            backtrack(finishTime, nextLate, newInDegree, nextAvailable, taskMap);
        }
    }
}